### PR TITLE
Moved serialization of `Cookie` to `Cookie`.

### PR DESCRIPTION
**Pull request by SamMousa**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
Breaks BC?    | yes/no
| Tests pass?   | yes
| Fixed issues  | #13135

### CODE REVIEW

**Comment by SamMousa** [comment]
Note that this also adds support for non-string values in cookies (when validation is enabled).

Should this be expanded so that cookies can contain any kind of data in all cases?
(Basically always serialize / unserialize regardless of validation?)

**Comment by samdark** [comment]
I think an exception for scalar values is made for cookies which are meant to be used via JavaScript.

**Comment by SamMousa** [comment]
> I think an exception for scalar values is made for cookies which are meant to be used via JavaScript.

What do you mean?

**Comment by samdark** [comment]
If the value is not an array, it's not encoded in any special way. That is for using it without decoding.

**Comment by SamMousa** [comment]
Are you sure?
https://github.com/yiisoft/yii2/blob/master/framework/web/Response.php#L397

I don't see that happening anywhere; as far as I can tell if cookie validation is enabled cookies get encoded.

**Comment by samdark** [comment]
You're right.

**Comment by SamMousa** [comment]
@samdark This one could be ready for merge, maybe. 8-)

**Comment by samdark** [comment]
I wonder how would that work with existing cookies set with data which was passed through old `serialize()/unserialize()`?

**Comment by samdark** [comment]
That's the main reason I was suggesting it for 2.1.x

**Comment by SamMousa** [comment]
Well those cookies would be come "invalid", meaning users lose whatever data is stored in there.
I don't feel it's a big issue, since users sessions tend to get remove whenever you do an upgrade anyway (that's what I do at least). And logging in again is not a big hassle.
Other than that, cookies that are used for preferences (like grid view settings) tend not to be validated anyway since they are changed via client side code.

Up to you which one you prefer, I'm in no rush to get this merged, just want it to be finished so it no longer needs a spot on my todo list.

**Comment by samdark** [comment]
Cookies may hold more than sessions.

File: framework/web/Cookie.php [hunk 1]
```diff
@@ -65,4 +67,38 @@ public function __toString()
...
+    /**
+     * Unserializes a cookie received by the client.
+     * @param string $data json encoded cookie value
+     * @return self|null
+     * @since 2.0.12
+     */
+    public static function fromDataString($data)
+    {
+        if (($unserialized = json_decode($data, true)) !== false) {
+            $result = new self();
```
Comment (samdark) [inline framework/web/Cookie.php]: How about `Yii::createObject()` so container settings are applies?

File: framework/web/Cookie.php [hunk 2]
```diff
@@ -65,4 +67,38 @@ public function __toString()
...
+
+    /**
+     * Unserializes a cookie received by the client.
+     * @param string $data json encoded cookie value
+     * @return self|null
+     * @since 2.0.12
+     */
+    public static function fromDataString($data)
+    {
+        if (($unserialized = json_decode($data, true)) !== false) {
```
Comment (samdark) [inline framework/web/Cookie.php]: I'd use `Json` helper since it produces much better errors.

File: framework/web/Cookie.php [hunk 3]
```diff
@@ -65,4 +67,38 @@ public function __toString()
...
+     * @return string the json encoded value of the cookie
+     * @throws InvalidConfigException if cookie does not have a name
+     * @since 2.0.12
+     */
+    public function toDataString()
+    {
+        if (!isset($this->name)) {
+            throw new InvalidConfigException("Cookie serialization requires a name.");
+        }
+        return json_encode([
```
Comment (samdark) [inline framework/web/Cookie.php]: I'd use `Json` helper since it produces much better errors.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/web/Cookie.php (hunk 2 above)
Comment (SamMousa) [inline framework/web/Cookie.php]: What should the behavior be regarding errors?
If I pass an invalid cookie, should the request object just crash (ie bubble up the exception) or should it silently ignore the cookie?

**Review by samdark — APPROVED** [review]

**Comment by klimov-paul** [comment]
Changing Cookie serialization from PHP `serialize()` to JSON can not be done at `2.0.x` as it causes a BC break: all existing Cookies in particular Yii-based will stop function after such update.

File: framework/web/Cookie.php [hunk 4]
```diff
@@ -65,4 +69,39 @@ public function __toString()
...
+     * @throws InvalidParamException if the data is not valid JSON
+     * @since 2.0.12
+     */
+    public static function fromDataString($data)
+    {
+        $unserialized = JSON::decode($data, true);
+        return \Yii::createObject([
+            'class' => get_called_class(),
+            'name' => $unserialized[0],
+            'value' => $unserialized[1],
```
Comment (klimov-paul) [inline framework/web/Cookie.php]: Keys `0` or `1` can be absent in unserialized data.
Such code may produce E_NOTICE.
It is better to perform `isset()` before like it was at original implementation.

**Review by klimov-paul — CHANGES_REQUESTED** [review]

**Comment by klimov-paul** [comment]
I do not like the idea: processing of the Cookie value is still depends on `Request` and `Response` since they  use `Security` to process the hash. 
It would be more consistent to keep all fucntionality related to hash composition at single place rather then extend it to the extra class.

File: framework/web/Cookie.php (hunk 4 above)
Comment (SamMousa) [inline framework/web/Cookie.php]: The data is validated first, you shouldn't pass unvalidated data into any function.

**Comment by SamMousa** [comment]
Well validation of user data is part of the request object, so that makes sense in my opinion.
Just like formatting the response is part of the response object.

**Comment by klimov-paul** [comment]
Note that all this 'cookie data string' support exists only because of necessity of cookie value protection, which is introduced and configured(!) at `Request`/`Response` level. Changes proposed here looks like encapsulation break to me.

**Comment by klimov-paul** [comment]
I can see absolutely no benefit in proposes changes - just a redundant overcomplication of code.

**Comment by SamMousa** [comment]
The serialization is not only for cookie protection, it also enables use of complex cookie values.

**Comment by klimov-paul** [comment]
> The serialization is not only for cookie protection, it also enables use of complex cookie values.

I can not see any complex valu handling in your PR. Besides what prevent us to simply use `Json` methods instead of PHP `serialize()`/`unserialize()` and `Request` and `Response`? Why there is a necessity of instroduction of extra methods at `Cookie` class?

**Comment by SamMousa** [comment]
Because it should be the cookies'  responsibility to handle its serialization / deserialization.
Discuss this with the team make a decision, I've got no more arguments to add.

**Comment by klimov-paul** [comment]
> Because it should be the cookies' responsibility to handle its serialization / deserialization.

It would be so in case I can use single `Cookie` class to get/set complex cookie value, like following:

```php
$cookie = Cookie::parse($_COOKIES['someComplexValue']);
var_dump($cookie->value);
```

but this code can not function since cookie value may contain security hash, which handled and configured at `Request` class.

In case there is no way to provide clear interface for cookie serialization/unserializion and single `Cookie` class value - why should we create a half-solution methods at `Cookie` class? Those will just confuse the developer, which expects it handles it all - while it does not take security hash into account.

Creating those methods just because 'it seems they should be there' does not seem right.
Either entire Cookie value processing including security hash should be moved inside the `Cookie` class - or it should remain intact.

**Comment by SamMousa** [comment]
Look at the tests..

It will work just fine like this:
````
$cookie = new Cookie();
$cookie->value = ['abc', 'def'];
$cookie->name = test;
````

This cookie can be serialized and unserialized by the request / response object, or other code.

**Comment by klimov-paul** [comment]
The raw value, which includes security hash, can not be parsed.

**Comment by SamMousa** [comment]
It only includes the security hash if it was signed by the Response. If you add it manually as you propose then it'll work without.

Your argument does not make sense. I could use it for any encryption / signing library, arguing that when encrypting data it does not work unless decrypted....
Of course if you sign the data and thus change it you have to undo that change, which is what the Request and Response objects do.

If you bypass them, the cookie serialization part will work just fine without it.

**Comment by klimov-paul** [comment]
>Your argument does not make sense. I could use it for any encryption / signing library, arguing that when encrypting data it does not work unless decrypted....

Cookies, which submitted per request are built and processed by `Request` class, which handles thier unserialization and decryption.
Standalone unserialization without decryption of the Cookie value make a little sence.
The only use case, in which I may want to unserialize some Cookie value on my own is getting this value from raw request, but in this case this value will contain the security hash in it.

Can you describe the use case, in which you get some random Cookie string value and want to unserialize it, but can be sure its value is secure and decrypted?

**Comment by SamMousa** [comment]
> Can you describe the use case, in which you get some random Cookie string value and want to unserialize it, but can be sure its value is secure and already decrypted?

When I use any other implementation than the default `Request` / `Response` class.

Note that "being sure" is not really relevant here. Can you be sure no developer will subclass `Response` but not `Request` and guarantee that they still use the same scheme for validation? (No you can not)

**Comment by klimov-paul** [comment]
> When I use any other implementation than the default Request / Response class. 

This is exactly the case - if you wish the Cookie ecryption/serialization change - you need to override `Request`/`Response` classes pair - this is the only consistent solution I see for such task. During such implementation another way for cookie serialization can be used as well. 

> Can you be sure no developer will subclass Response but not Request and guarantee that they still use the same scheme for validation? 

When it comes to override existing core functionality developer is responsible for its results. I can not see how new methods in `Cookie` class ensures anything in this field - who can garantee developer will use those methods, while creating its own Cookie protection solution?

Yii is not a Symfony and it does not follow SOLID methodology - we do not introduce a new abstraction layer just because we have found a potential place for it.

**Comment by sergeymakinen** [comment]
I think it’s useful to extract a `Cookie` serialization completely, so it can be swapped without the need to touch `Request`/`Response` (like using JWT for signing/validation).

**Comment by samdark** [comment]
I think in this case serialization should be extracted as @sergeymakinen suggested.

**Comment by SamMousa** [comment]
Signing and validation will require configuration though.
Currently that is set in the `Request` component (`cookieValidationKey`).

I think serialization (transformation from complex to string) is a separate step from signing (transformation from string to protected string).
I agree that signing should be removed from `Request` / `Response`, but it should not be in `Cookie`; since cookies are not the only thing you can / could sign, you could sign any piece of data that you want to send to the client and get back in a later request. (For example, hidden form field).

That said, this is a first step to separate the serialization part, feel free to make the next step and make a PR for it...

This PR does what I want it to do, if you don't want that in the framework that's fine, close this issue. If you do want it but see improvements for the future I suggest we create a separate issue for that and mark this as "ready to merge into 2.1".

**Comment by samdark** [comment]
As @klimov-paul noted, that's significant enough to go to 2.1. I like these changes overall so would not vote for not accepting the pull request.

**Comment by samdark** [comment]
Won't be merged into 2.0.

**Pull request closed without merging**